all_seaing_autonomy/all_seaing_autonomy/roboboat/WaypointFinder.py

```diff
@@ -0,0 +1,221 @@
+        return True
+    
+    def ccw(self, a, b, c):
+        """Check if the points a, b, c are counterclockwise, in this order, and return True, otherwise return False"""
+        # literally just shoelace formula
+        area = a[0]*b[1] + b[0]*c[1] + c[0]*a[1] - a[1]*b[0] - b[1]*c[0] - c[1]*a[0]
+        if area > 0:
+            return True
+        else:
+            return False
+    
+    # euler <-> quaternion transformation code from: https://gist.github.com/salmagro/2e698ad4fbf9dae40244769c5ab74434
+    def euler_from_quaternion(self, quaternion):
+        """
+        Converts quaternion (w in last place) to euler roll, pitch, yaw
+        quaternion = [x, y, z, w]
+        Bellow should be replaced when porting for ROS 2 Python tf_conversions is done.
+        """
+        x = quaternion.x
+        y = quaternion.y
+        z = quaternion.z
+        w = quaternion.w
+
+        sinr_cosp = 2 * (w * x + y * z)
+        cosr_cosp = 1 - 2 * (x * x + y * y)
+        roll = math.atan2(sinr_cosp, cosr_cosp)
+
+        sinp = 2 * (w * y - z * x)
+        pitch = math.asin(sinp)
+
+        siny_cosp = 2 * (w * z + x * y)
+        cosy_cosp = 1 - 2 * (y * y + z * z)
+        yaw = math.atan(siny_cosp, cosy_cosp)
+
+        return roll, pitch, yaw
+    
+    def quaternion_from_euler(self, roll, pitch, yaw):
+        """
+        Converts euler roll, pitch, yaw to quaternion (w in last place)
+        quat = [x, y, z, w]
+        Bellow should be replaced when porting for ROS 2 Python tf_conversions is done.
+        """
+        cy = math.cos(yaw * 0.5)
+        sy = math.sin(yaw * 0.5)
+        cp = math.cos(pitch * 0.5)
+        sp = math.sin(pitch * 0.5)
+        cr = math.cos(roll * 0.5)
+        sr = math.sin(roll * 0.5)
+
+        q = [0] * 4
+        q[0] = cy * cp * cr + sy * sp * sr
+        q[1] = cy * cp * sr - sy * sp * cr
+        q[2] = sy * cp * sr + cy * sp * cr
+        q[3] = sy * cp * cr - cy * sp * sr
+
+        return q
+    
+    def filter_front_buoys(self, pair, buoys):
+        # (red, green)
+        return [buoy for buoy in buoys if self.ccw(pair[0], pair[1], buoy)]
+
+    def next_pair(self, prev_pair, red, green):
+        """Find the next buoy pair (red left, green right) from the previous pair,
+        by checking the closes one to the middle of the previous buoy pair that's in front of the pair"""
+        return (self.get_closest_to(self.midpoint_pair(prev_pair), self.filter_front_buoys(prev_pair, red)), self.get_closest_to(self.midpoint_pair(prev_pair), self.filter_front_buoys(prev_pair, green)))
+    
+    def pair_to_pose(self, pair):
+        return Pose(position = Point(x = pair[0], y = pair[1]))
+    
+    def pair_angle_to_pose(self, pair, angle):
+        quat = self.quaternion_from_euler(0,angle,0)
+        return Pose(position = Point(x = pair[0], y = pair[1]), orientation = Quaternion(x = quat[0], y = quat[2], z = quat[2], w = quat[3]))
+
+    def generate_waypoints(self):
```

Comment:
(discussed irl) would be good to include some documentation on how this works, either in docstring or in the wiki